File: c++/max-sum-contiguous-subarray.cpp
/*
task

Find the contiguous subarray within an array, A of length N which has the largest sum.

https://www.interviewbit.com/problems/max-sum-contiguous-subarray/

Example Input
A = [-2, 1, -3, 4, -1, 2, 1, -5, 4]

Output
6

Explanation
The subarray [4, -1, 2, 1] has the maximum possible sum of 6.

*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
    public:
    int maxSubArray(const vector<int>&);
};

int Solution::maxSubArray(const vector<int> &A)
{
    // hash map between vector and value
    int currentSum = A[0];
    int maxSum = currentSum;

    for(int idx = 1; idx < A.size(); idx++)
    {
        // currentSum = max(A[idx], currentSum + A[idx])
        if(A[idx] >= currentSum + A[idx])
        {
            currentSum = A[idx];
        }
        else
        {
            currentSum += A[idx];
        }

        // maxSum = max(maxSum, currentSum)
        if(currentSum > maxSum)
        {
            maxSum = currentSum;
        }
    }

    return maxSum;
}

int main(int argc, char* argv[])
{

    Solution solution;
    
    vector<int> v1({1, 2, 3, 4, -10});
    vector<int> v2({-2, 1, -3, 4, -1, 2, 1, -5, 4});
    vector<int> v3({20, 1, 1, 1, 1, 1, -400, 1000, 0, -500, 0});

    cout << to_string(solution.maxSubArray(v1)) << endl;
    cout << to_string(solution.maxSubArray(v2)) << endl;
    cout << to_string(solution.maxSubArray(v3)) << endl;

    return EXIT_SUCCESS;
}

File: python/max-sum-contiguous-subarray.py
'''
task

Find the contiguous subarray within an array, A of length N which has the largest sum.

https://www.interviewbit.com/problems/max-sum-contiguous-subarray/

Example Input
A = [-2, 1, -3, 4, -1, 2, 1, -5, 4]

Output
6

Explanation
The subarray [4, -1, 2, 1] has the maximum possible sum of 6.

'''

class Solution:
    # @param A : tuple of integers
    # @return an integer
    def maxSubArray(self, A : tuple):

        current_sum = A[0]
        max_sum = current_sum

        for idx in range(1, len(A), 1):
            current_sum = max(A[idx], current_sum + A[idx])
            max_sum = max(current_sum, max_sum)

        return max_sum

s = Solution()
result = s.maxSubArray(A = [1, 2, 3, 4, -10]) # returns 10
result = s.maxSubArray(A = [-2, 1, -3, 4, -1, 2, 1, -5, 4]) # returns 6
result = s.maxSubArray(A = [20, 1, 1, 1, 1, 1, -400, 1000, 0, -500, 0]) # returns 1000
